test/test_pipeline/components/setup/test_setup_networks.py

```diff
@@ -31,6 +31,9 @@ def test_pipeline_fit(self, fit_dictionary_tabular, embedding, backbone, head):
         """This test makes sure that the pipeline is able to fit
         every combination of network embedding, backbone, head"""
 
+        # increase number of epochs to test for performance
+        fit_dictionary_tabular['epochs'] = 50
```

Comment:
Can we remove it or can we use NamedTuple instead? Because it is already defined in the contest and we do not want to change the variable locally. Especially, I am afraid of modifying the value without being recognized by anyone. If it is not appropriate change for this PR, I will raise an issue.

Reply:
I didnt get the problem here. Could you explain again? For certain tests we want to check the performance of the pipelines, and hence we want higher number of epochs. Where do you want to use a NamedTuple?

Reply:
There are mainly two reasons:
1. it cannot reflect the changes directly in `fit_dictionary_tabular` in conftest
2. some people change `fit_dictionary_tabular` in conftest without noticing the local modifications

if we can modify the configuration, i.e. in this case `fit_dictionary_tabular`, when some people just thought that once we change the `fit_dictionary_tabular` in conftest, it will be reflected to all the functions, they suffer from these local modifications. Since dict can be modified easily, it is not secured with respect to this matter and it potentially increases our works in the future.

Reply:
I think the changes to `fit_dictionary_tabular` in any local test are based on a need for that test. Therefore, if in a test we notice something should be changed, it can be done so using a dictionary and for me, it does not cause any issues. Using a NamedTuple for `fit_dictionary` will cause the following issues:-
1. As tuples are immutable, to make a single change we'll have to change the whole tuple instance.
2. `fit dictionary` is anyways a `Dict`that is used throughout the pipeline so we'll have to always change the namedTuple to a dictionary for each test. 
3. If we still want a named tuple for `fit_dictionary`, that is a large refactor to the whole AutoPyTorch, which I don't think we need. Moreover, we are moving to fit_params API of sklearn as soon as possible anyways.

Reply:
Ah, I did not know that `fit_dictionary` is used in the main parts.
OK, then let's not change this time. But can we define this epoch inside the function to fetch this dictionary?
Currently the function `get_fit_dictinary` in `conftest.py` does not support kwargs modifications, so can we enable to pass the variables to modify?
```
def get_fit_dictionary(X, y, validator, backend, **kwargs):
    ...
    fit_dictionary = {
        ....
    }
    
    if set(kwargs.keys()) - set(fit_dictionary.keys()) >= 1:
        raise ValueError('Unknown keys were provided.')

    fit_dicitonary.update(kwargs)
    ...
    return fit_dictionary
```

Just in case, I will leave some comments here and probably post on the Issue page.
First, I checked where it is used and there were only 5 lines which use `fit_dictionary`.
- `abstract_evaluator.py` and `train_evaluator.py`
According to the implementations, you are right and we need to update the dictionary sometimes.
So we cannot change it to tuple.

Reply:
I do not see why it is going to be a lot of work.
Because **kwargs is not a mandatory argument and you do not need to change any other codes where you do not need to change the epoch.
It means you need to change the function and the places where you need to change the epoch.
If I count correctly, it will be only 4- or 5-lines modifications.
Am I wrong here? Maybe I am missing some points.

About this time, you do not need this change, because test codes are not complicated compared to the main parts. 

========================== REVIEW part =========================-
However, It is more like general advice and since it is quite easy to change this time, I do not see any reasons why we do not change.
You know the variable range, right?
We should not change variables in wider ranges. 
When some issues happen, it is quite hard to know where the bugs are coming from.
In fact, it happened when I was inspecting the memory issues.
That is why I assume it is a common bad practice in this project and I need to persistently mention to condition what and why your codes matter when I find such bad practices.

Reply:
Firstly, it may be a simple modification. But I don't like the changes we'll have to make to make this work. To explain further, for every function that needs this epoch value to be changed, we'll have to pass the epochs in the pytest.mark.parameterize function  `@pytest.mark.parametrize("fit_dictionary_tabular_dummy", [("classification", 50)], indirect=True)`, see [here](https://github.com/ravinkohli/Auto-PyTorch/blob/44dec19a9f00ff2eb3618d2acaa2bd4691758ad5/test/test_pipeline/test_tabular_classification.py#L428). After this we'll have to change the function call [here](https://github.com/ravinkohli/Auto-PyTorch/blob/44dec19a9f00ff2eb3618d2acaa2bd4691758ad5/test/conftest.py#L315) to accept a tuple and use the first one of the tuple for checking if its classification or not and the second part to send the epochs to the `get_fit_dictionary` where we'll also have to check if epochs were given or not. I don't think all this is necessary just for a test. 
Secondly, I don't see how this will make memory debugging process any more complicated. Please let me know if you want to. 
Thirdly, for changing this dictionary to tuple, as I previously wrote, we plan on making a refactor that will get rid of this fit_dictionary entirely. We'll instead be using fit_params of sklearn, which also means that changing fit dictionary to a tuple is moot. 
Finally, as for the "bad practices in this project", we have tried to ensure that we have proper checks for coding style and static type checking, and where we felt necessary, use namedTuples or child classes of that instead of using dictionaries. If you still find places that are not optimal, please raise an issue about this and we'll try to resolve it.